Implemented #404.

toddATavail opened the pull request:
- Added --showSize to get-bucket. Introduces two new fields, fileCount and totalSize, that represent the count of files and the aggregate size of all files, respectively.
- Omission of --showSize results in the original behavior.
- Added venv to .gitignore.

Additional documentation is available in a comment on #404.

bwbeach commented on b2/console_tool.py:
```diff
@@ -420,21 +422,42 @@ def run(self, args):
+                    return 0
+                else:
+                    files = list(b.ls("", args.bucketName))
+                    # Note that `files` is a list of tuples of (file_version_info, folder_name). We
```
Nice comment.  This makes the code very clear.

bwbeach commented on b2/console_tool.py:
```diff
@@ -420,21 +422,42 @@ def run(self, args):
+                    self._print(json.dumps(b.bucket_dict, indent=4, sort_keys=True))
+                    return 0
+                else:
+                    files = list(b.ls("", args.bucketName))
```
For large buckets, it would be better not to build the entire list of files.  `ls` yields the results one at a time.  If you process them one at a time without building a list, it won't require so much memory.

Right now, there are buckets in B2 with almost a billion files.

toddATavail replied:
I was wondering about this while I was coding it. I recognized that it wouldn't scale for very large buckets, but didn't have much idea about the scalability of the system as a whole in any case. I'll rewrite to accumulate without interpolating the entire generator.

bwbeach commented:
The travis failure is because of code formatting.  This project requires that the formatting be what `yapf` would do.  (It's not always what I prefer, but at least it's consistent.)  If you run the `pre-commit.sh` script, it should fix up the formatting.

bwbeach commented:
Looks good. One final issue to think about: What does "bucket size" mean?

B2 supports [file versioning and hiding](https://www.backblaze.com/b2/docs/file_versions.html). What should the answer be if there are multiple versions of a file, or if a file is hidden but still using space?

`Bucket.ls` has options for showing all file versions.

Also, in the doc string, it's probably worth mentioning that `--showSize` will list the files, and incur transaction costs for doing so. That will avoid having customers call support when they see a bill for lots of calls to list files.

toddATavail commented:
The original poster of #404 asked for equivalent behavior to what the website provides. This is, therefore, how I implemented it. I treat versions as distinct files and I count the existence of a hidden file toward the membership count of the bucket, but I do not include the size of hidden files in the aggregate size of the bucket. The answer produced by `--showSize` agrees with the answer provided on the *"B2 Cloud Storage Buckets"* page of the website.

It is a good idea to clarify in the documentation that there are additional transactions involved with `--showSize`, so that customers understand the financial implications of this option. I'll take care of that.

Let me know if `--showSize` should have different behavior, take additional arguments to fine tune behavior, etc.

bwbeach commented:
The doc changes look good.

I think the best place to put tests for the new functionality is in `test_console_tool.py`.  It runs against a B2 simulator, and will let you exercise the new code.

bwbeach commented:
I think there may be bugs in the web site showing total bucket size.  We'll need to look into that.

The answer from the new code looks good for multi-version files.  I hadn't expected it to be right, because it's not passing `showVersions=True` to `ls`.  I need to understand why that is.

This looks like the right answer:

```
(p35) stone:B2_Command_Line_Tool brianb$ python -m b2 get-bucket --showSize bwb-test
{
    "accountId": "890dc49cef59",
    "bucketId": "28a9106d0cc4d91c5e1f0519",
    "bucketInfo": {},
    "bucketName": "bwb-test",
    "bucketType": "allPrivate",
    "corsRules": [],
    "fileCount": 3,
    "lifecycleRules": [
        {
            "daysFromHidingToDeleting": 1,
            "daysFromUploadingToHiding": null,
            "fileNamePrefix": ""
        }
    ],
    "revision": 2,
    "totalSize": 7
}
(p35) stone:B2_Command_Line_Tool brianb$ python -m b2 ls --long bwb-test
4_z28a9106d0cc4d91c5e1f0519_f1119c6eb78d6fd97_d20180316_m200016_c000_v0001002_t0030  upload  2018-03-16  20:00:16          4  a
4_z28a9106d0cc4d91c5e1f0519_f1124c422d541f229_d20180316_m195932_c000_v0001045_t0042  upload  2018-03-16  19:59:32          2  b
(p35) stone:B2_Command_Line_Tool brianb$ python -m b2 ls --long --versions bwb-test
4_z28a9106d0cc4d91c5e1f0519_f1119c6eb78d6fd97_d20180316_m200016_c000_v0001002_t0030  upload  2018-03-16  20:00:16          4  a
4_z28a9106d0cc4d91c5e1f0519_f119f896de242fabd_d20180316_m195932_c000_v0001058_t0027  upload  2018-03-16  19:59:32          1  a
4_z28a9106d0cc4d91c5e1f0519_f1124c422d541f229_d20180316_m195932_c000_v0001045_t0042  upload  2018-03-16  19:59:32          2  b
```

bwbeach commented:
The behavior of `Bucket.ls` is weird.  Weird enough that I'd call it a bug and would like to clean it up someday.  It doesn't descend into folders; when it encounters one it returns the first file in the folder, along with a folder name.  This makes the bucket size wrong.  Passing in `recursive=True` should fix it.

```
(p35) stone:B2_Command_Line_Tool brianb$ python -m b2 ls --long --versions bwb-test 
4_z28a9106d0cc4d91c5e1f0519_f1119c6eb78d6fd97_d20180316_m200016_c000_v0001002_t0030  upload  2018-03-16  20:00:16          4  a
4_z28a9106d0cc4d91c5e1f0519_f119f896de242fabd_d20180316_m195932_c000_v0001058_t0027  upload  2018-03-16  19:59:32          1  a
4_z28a9106d0cc4d91c5e1f0519_f1124c422d541f229_d20180316_m195932_c000_v0001045_t0042  upload  2018-03-16  19:59:32          2  b
                                                                                  -       -           -         -          0  c/
(p35) stone:B2_Command_Line_Tool brianb$ python -m b2 ls --long --versions bwb-test c
4_z28a9106d0cc4d91c5e1f0519_f119323d8fe2f8ac3_d20180316_m200730_c000_v0001060_t0050  upload  2018-03-16  20:07:30          8  c/c1
4_z28a9106d0cc4d91c5e1f0519_f108ce6b551ce0261_d20180316_m200730_c000_v0001024_t0042  upload  2018-03-16  20:07:30         16  c/c2
(p35) stone:B2_Command_Line_Tool brianb$ python -m b2 get-bucket --showSize bwb-test
{
    "accountId": "890dc49cef59",
    "bucketId": "28a9106d0cc4d91c5e1f0519",
    "bucketInfo": {},
    "bucketName": "bwb-test",
    "bucketType": "allPrivate",
    "corsRules": [],
    "fileCount": 4,
    "lifecycleRules": [
        {
            "daysFromHidingToDeleting": 1,
            "daysFromUploadingToHiding": null,
            "fileNamePrefix": ""
        }
    ],
    "revision": 2,
    "totalSize": 15
}
```

toddATavail commented:
I have addressed your suggestions, Brian. I passed the mentioned flags to `Bucket.ls()` and I added two unit tests, one that tests against an empty bucket and one that tests against a bucket containing a standard test file. The unit tests all passed for me, but it looks like (as of this comment) that the CI tests haven't finished running.

bwbeach commented:
Looks good.  Thanks!

I'll merge once the build is done.
